runtime/compare.c

```diff
@@ -109,12 +109,45 @@ static intnat compare_val(value v1, value v2, int total)
 #define GREATER 1
 #define UNORDERED ((intnat)1 << (8 * sizeof(value) - 1))
 
+static bool compare_poll_actions(struct compare_item** sp,
+                                 struct compare_stack* stk,
+                                 value *orig1, value *orig2,
+                                 value *v1, value *v2)
+{
+  value o1 = *orig1, o2 = *orig2;
+  if (caml_check_pending_actions()) {
+    compare_free_stack(stk);
+    CAMLparam2(o1, o2);
+    value exn = caml_do_pending_actions_exn();
+    if (Is_exception_result(exn)) {
+      caml_raise(exn);
+    }
+    CAMLdrop;
+    *v1 = *orig1 = o1;
+    *v2 = *orig2 = o2;
+    *sp = stk->stack = stk->init_stack;
+    return true;
+  }
+  return false;
+}
+
 /* The return value of compare_val is as follows:
       > 0                 v1 is greater than v2
       0                   v1 is equal to v2
       < 0 and > UNORDERED v1 is less than v2
       UNORDERED           v1 and v2 cannot be compared */
 
+#define COMPARE_DO_POLL()                                           \
+  do {                                                              \
+    signal_poll_timer = 0;                                          \
+    if (compare_poll_actions(&sp, stk, &orig1, &orig2, &v1, &v2))   \
+      goto loop_start;                                              \
+  } while (0)
```

Comment:
This could be just the source of the `poll:` label, with COMPARE_INC_POLL_TIMER doing a `goto loop`, right?

Reply:
(Also: I'm not too excited by macros being defined at the toplevel that contains bound occurrences of names that are not currently in scope. I made suggestions that I think would let us get rid of `COMPARE_DO_POLL`, and I would rather write the code corresponding to `COMPARE_IN_POLL_TIMER` directly at each callsite, or possibly defining the macro within the body of the `do_compare_val` function.)

Reply:
> This could be just the source of the `poll:` label, with COMPARE_INC_POLL_TIMER doing a `goto loop`, right?

That's not quite true, we need it to return from where it left off if the poll finds nothing when comparing strings or double arrays. I could make `COMPARE_INC_POLL_TIMER` expand to an expression instead, and add `goto loop_start`s and `continue`s explicitly at the call site to make intent clearer:
```c
static intnat do_compare_val(/*...*/) {
#define COMPARE_INC_POLL_TIMER(inc)                                     \
  (((signal_poll_timer += (inc)) >= COMPARE_SIGNAL_POLL_PERIOD) &&      \
   (signal_poll_timer = 0,                                              \
    compare_poll_actions(&sp, stk, &orig1, &orig2, &v1, &v2)))
// ...
  if (COMPARE_INC_POLL_TIMER(1)) goto loop_start;
// ...
}
```
Or of course inline it at call sites, at the cost of some code duplication.

Reply:
Ah of course, I was missing the fact that we return normally in absence of pending actions. Sorry for the noise.